### PR TITLE
add support for field definition directives, handlers, and identity

### PR DESCRIPTION
**Pull request by michaelraines**
https://app.asana.com/0/1182291590501927/1205474699889736/f

This change implements SOME support for directives in our GQL library and adds hooks to handle directives at runtime. It also adds a gen time assertion regarding allow identity usage.

Field Def Directive/Handler Support:
image
image

Gen time checks: (This is a bit of a hack putting it right into the generator, but we are moving away from this soon anyways)
image

Current "Hacks":
* How the directive assertion system (making sure all relevant fields are explicitly annotated) is a part of the generator itself rather than having the generator expose some sort of API/Config hooks to wire that in.

* The rendering of ast.Value currently only supports Boolean arguments in directives. There are lots of other types that match that interface, but none of our directives currently use them.

### CODE REVIEW

File: cmd/graphql2go/main.go
Comment (samuel) [inline cmd/graphql2go/main.go]:
```suggestion
			g.failf("@allowAssumedIdentity directive required on field %q since parent %q is top level or has @allowAssumedIdentity(allow: true)", f.Name.Value, parentName)
```

File: cmd/graphql2go/main.go
Comment (samuel) [inline cmd/graphql2go/main.go]:
```suggestion
			g.failf("unable to resolve def for type %q", f.Type)
```

**Comment by kunaljham** [comment]
@michaelraines in what way is this a hack? what are the limitations here? I ask because we just made the decision to pause the work on the new gqlllib, so we will likely have to live with this implementation and usage in our old gqllib for a while.

https://spruce.slack.com/archives/C03A6TQEZ/p1697065329335249

**Comment by michaelraines** [comment]
@kunaljham the two main callouts I would make as "hacks" are 

* How the directive assertion system (making sure all relevant fields are explicitly annotated) is a part of the generator itself rather than having the generator expose some sort of API/Config hooks to wire that in.
* The rendering of ast.Value currently only supports Boolean arguments in directives. There are lots of other types that match that interface, but none of our directives currently use them.

**Review by kunaljham — APPROVED** [review]
looks good to me, but worth it for @samuel to review and have the final say on whether or not things look good.

**Comment by michaelraines** [comment]
Merging but please still let me know if there is feedback @samuel

**Review by samuel — COMMENTED** [review]
Looks fine to me. I think the main thing is to make sure that how the directives are defined in the spec are standard graphql so that we don't have any issues with Apollo or the new gql lib once it's merged. To be clear, I mean the schema definition and how it's implemented in the code doesn't matter as much since it's easier to change.

**Pull request merged**
